### PR TITLE
Automatically restage block volumes if necessary

### PR DESCRIPTION
**Pull request by ejweber**
#### Which issue(s) this PR fixes:

longhorn/longhorn#8009

#### What this PR does / why we need it:

If `NodePublishVolume` is called when a block volume is not staged (see context in longhorn/longhorn#8009), we automatically call `NodeUnstageVolume` and `NodeRestageVolume`). We already had this behavior for mount volumes.

### CODE REVIEW

File: csi/node_server.go [hunk 1]
```diff
@@ -207,6 +177,24 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
 
+	if isBlock {
+		devicePath := getStageBlockVolumePath(stagingTargetPath, volumeID)
+		_, err := os.Stat(devicePath)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return nil, status.Errorf(codes.Internal, errors.Wrapf(err, "failed to stat device %s", devicePath).Error())
+			}
+		}
+
```
Comment (ejweber) [inline csi/node_server.go:188]: I removed the endpoint fallback like:

```
// Fall back to the controller endpoint if the device path under the stagingTargetPath doesn't exist
log.Infof("Device path %s doesn't exist, falling back to controller endpoint %s", devicePath, volume.Controllers[0].Endpoint)
devicePath = volume.Controllers[0].Endpoint
```

This was necessary before because `NodeStageVolume` was a no-op in < `v1.6.0`, so a volume might not be staged at `stagingTargetPath` before `NodePublishVolume` is called. The new restaging logic will ensure all `NodePublishVolume` commands succeed (with automatic restaging if necessary). Even if a volume was previously staged (via no-op) in < `v1.6.0`, calling `NodeUnstageVolume` and `NodeRestageVolume` is safe to do. See the new comment in `restageRequired` for additional details.

**Comment by ejweber** [comment]
To test:

Follow steps in https://github.com/longhorn/longhorn/issues/8009#issue-2149896606.

5. When the node comes back, the newly created pod is able to get running again. (As seen in the output below, In my cluster, most of the time there is actually a pod that becomes running and is quickly terminated before another becomes running. This is a race between the Longhorn automatic remount mechanism and the newly started node. After investigating, I am confident it has no bearing on this issue.)

```
eweber@laptop:~/longhorn> k get pod -owide --output-watch-events -w
EVENT      NAME                                 READY   STATUS    RESTARTS   AGE    IP           NODE                                NOMINATED NODE   READINESS GATES
ADDED      block-volume-test-5b48dc8675-mctg9   1/1     Running   0          114s   10.42.5.21   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-mctg9   1/1     Terminating   0          2m12s   10.42.5.21   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
DELETED    block-volume-test-5b48dc8675-mctg9   1/1     Terminating   0          2m12s   10.42.5.21   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
ADDED      block-volume-test-5b48dc8675-4dfgn   0/1     Pending       0          0s      <none>       <none>                              <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   0/1     Pending       0          0s      <none>       eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   0/1     ContainerCreating   0          19s     <none>       eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   1/1     Running             0          43s     10.42.5.41   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   1/1     Terminating         0          43s     10.42.5.41   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
ADDED      block-volume-test-5b48dc8675-8tkt6   0/1     Pending             0          0s      <none>       <none>                              <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-8tkt6   0/1     Pending             0          0s      <none>       eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-8tkt6   0/1     ContainerCreating   0          0s      <none>       eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-8tkt6   1/1     Running             0          2s      10.42.5.42   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   0/1     Terminating         0          45s     <none>       eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   0/1     Terminating         0          46s     10.42.5.41   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
MODIFIED   block-volume-test-5b48dc8675-4dfgn   0/1     Terminating         0          46s     10.42.5.41   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
DELETED    block-volume-test-5b48dc8675-4dfgn   0/1     Terminating         0          46s     10.42.5.41   eweber-v126-worker-9c1451b4-w6977   <none>           <none>
```

6. The CSI plugin logs show `NodeStageVolume` was called by the Longhorn CSI plugin itself while serving a `NodePublishVolume` request.

```
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="NodePublishVolume is called with req volume_id:\"pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95\" staging_target_path:\"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95\" target_path:\"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/9e8663e2-aeda-4023-94d4-1b242fb6619f\" volume_capability:<block:<> access_mode:<mode:SINGLE_NODE_WRITER > > volume_context:<key:\"csi.storage.k8s.io/ephemeral\" value:\"false\" > volume_context:<key:\"csi.storage.k8s.io/pod.name\" value:\"block-volume-test-5b48dc8675-4dfgn\" > volume_context:<key:\"csi.storage.k8s.io/pod.namespace\" value:\"default\" > volume_context:<key:\"csi.storage.k8s.io/pod.uid\" value:\"9e8663e2-aeda-4023-94d4-1b242fb6619f\" > volume_context:<key:\"csi.storage.k8s.io/serviceAccount.name\" value:\"default\" > volume_context:<key:\"dataLocality\" value:\"disabled\" > volume_context:<key:\"fromBackup\" value:\"\" > volume_context:<key:\"fsType\" value:\"ext4\" > volume_context:<key:\"numberOfReplicas\" value:\"3\" > volume_context:<key:\"staleReplicaTimeout\" value:\"30\" > volume_context:<key:\"storage.kubernetes.io/csiProvisionerIdentity\" value:\"1708717921145-2601-driver.longhorn.io\" > volume_context:<key:\"unmapMarkSnapChainRemoved\" value:\"ignored\" > " func="csi.(*NodeServer).NodePublishVolume" file="node_server.go:82" component=csi-node-server function=NodePublishVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=error msg="Staging target path /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 is no longer valid for volume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func="csi.(*NodeServer).NodePublishVolume" file="node_server.go:157" component=csi-node-server error="<nil>" function=NodePublishVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=warning msg="Calling NodeUnstageVolume for volume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func="csi.(*NodeServer).NodePublishVolume" file="node_server.go:159" component=csi-node-server function=NodePublishVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="NodeUnstageVolume is called with req volume_id:\"pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95\" staging_target_path:\"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95\" " func="csi.(*NodeServer).NodeUnstageVolume" file="node_server.go:541" component=csi-node-server function=NodeUnstageVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Trying to force unmount potential mount point /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func=csi.unmount file="util.go:349"
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Trying to cleanup mount point /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func=csi.cleanupMountPoint file="util.go:371"
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Trying to force unmount potential mount point /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func=csi.unmount file="util.go:349"
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Cleaned up mount point /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func=csi.cleanupMountPoint file="util.go:377"
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=warning msg="Device /dev/mapper/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 is not an active LUKS device" func=crypto.DeviceEncryptionStatus file="crypto.go:171" error="failed to execute: /usr/bin/nsenter [nsenter --mount=/proc/4270/ns/mnt --ipc=/proc/4270/ns/ipc cryptsetup status pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95], output /dev/mapper/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 is inactive.\n, stderr : exit status 4"
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Volume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 unmounted from node path /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func="csi.(*NodeServer).NodeUnstageVolume" file="node_server.go:590" component=csi-node-server function=NodeUnstageVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=warning msg="Calling NodeStageVolume for volume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func="csi.(*NodeServer).NodePublishVolume" file="node_server.go:165" component=csi-node-server function=NodePublishVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="NodeStageVolume is called with req volume_id:\"pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95\" staging_target_path:\"/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95\" volume_capability:<block:<> access_mode:<mode:SINGLE_NODE_WRITER > > volume_context:<key:\"csi.storage.k8s.io/ephemeral\" value:\"false\" > volume_context:<key:\"csi.storage.k8s.io/pod.name\" value:\"block-volume-test-5b48dc8675-4dfgn\" > volume_context:<key:\"csi.storage.k8s.io/pod.namespace\" value:\"default\" > volume_context:<key:\"csi.storage.k8s.io/pod.uid\" value:\"9e8663e2-aeda-4023-94d4-1b242fb6619f\" > volume_context:<key:\"csi.storage.k8s.io/serviceAccount.name\" value:\"default\" > volume_context:<key:\"dataLocality\" value:\"disabled\" > volume_context:<key:\"fromBackup\" value:\"\" > volume_context:<key:\"fsType\" value:\"ext4\" > volume_context:<key:\"numberOfReplicas\" value:\"3\" > volume_context:<key:\"staleReplicaTimeout\" value:\"30\" > volume_context:<key:\"storage.kubernetes.io/csiProvisionerIdentity\" value:\"1708717921145-2601-driver.longhorn.io\" > volume_context:<key:\"unmapMarkSnapChainRemoved\" value:\"ignored\" > " func="csi.(*NodeServer).NodeStageVolume" file="node_server.go:364" component=csi-node-server function=NodeStageVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Volume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 device /dev/longhorn/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 contains filesystem of format " func="csi.(*NodeServer).NodeStageVolume" file="node_server.go:450" component=csi-node-server function=NodeStageVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Bind mounting device /dev/longhorn/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 at /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func="csi.(*NodeServer).nodePublishBlockVolume" file="node_server.go:328" component=csi-node-server function=nodePublishBlockVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Volume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 device /dev/longhorn/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 available for usage as block device" func="csi.(*NodeServer).NodeStageVolume" file="node_server.go:493"
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Bind mounting device /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/staging/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95 at /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95/9e8663e2-aeda-4023-94d4-1b242fb6619f" func="csi.(*NodeServer).nodePublishBlockVolume" file="node_server.go:328" component=csi-node-server function=nodePublishBlockVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="Published BlockVolume pvc-cd3eb0e3-b7e3-4b4a-9519-51fa8a8e7e95" func="csi.(*NodeServer).NodePublishVolume" file="node_server.go:194" component=csi-node-server function=NodePublishVolume
[longhorn-csi-plugin-7q8l4] time="2024-02-23T19:57:45Z" level=info msg="NodePublishVolume: rsp: {}" func=csi.logGRPC file="server.go:141"
```

**Review by PhanLe1010 — APPROVED** [review]
It LGTM 

A quick question, I can understand how this PR make the new pod start but how does it PR resolve the issue that old pod is stuck in terminating? https://github.com/longhorn/longhorn/issues/8009#:~:text=The%20situation%20appears,directory%2C%20it%20cannot.

File: csi/node_server.go [hunk 2]
```diff
@@ -147,45 +147,15 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
...
-	}
-	if unstageVolumeRequired {
+	// It may be necessary to restage the volume before we can publish it. For example, sometimes kubelet calls
+	// NodePublishVolume without calling NodeStageVolume. According to the CSI spec, we should be able to respond with
+	// FailedPrecondition and expect kubelet to call NodeStageVolume again, but as of Kubernetes v1.27 it does not.
+	isBlock := volumeCapability.GetBlock() != nil
+	restageRequired, err := restageRequired(volume, volumeID, stagingTargetPath, mounter, isBlock)
+	if restageRequired {
 		msg := fmt.Sprintf("Staging target path %v is no longer valid for volume %v", stagingTargetPath, volumeID)
 		log.WithError(err).Error(msg)
```
Comment (james-munson) [inline csi/node_server.go:157]: It seems reasonable that this is not really an Error level log, but just Warning like the one immediately following it.  Is it possible to all log.WithError(err).Warning(msg) or failing that, to format the `err` explanation from restageRequired() into `msg`?

**Review by james-munson — COMMENTED** [review]
LGTM.  Just one nit.

**Comment by ejweber** [comment]
> A quick question, I can understand how this PR make the new pod start but how does this PR resolve the issue that old pod is stuck in terminating? [longhorn/longhorn#8009](https://github.com/longhorn/longhorn/issues/8009)

This is a good question. I didn't understand it at the time I wrote up the issue, but the pods that are stuck in terminating are NOT the ones that existed before the node reboot. In each support bundle, we see this type of pod cleaned up successfully after the CSI driver becomes available. Rather, they are created AFTER the node reboot in the broken state described and then subsequently killed. The common reason I see in my own testing is Longhorn's remount logic. So:

- The responsible controller creates a replacement `pod_replacement` while the node is down.
- `pod_replacement` transitions to starting as seen as the node is back.
- `pod_replacement` is broken because `NodePublishVolume` is called without `NodeStageVolume`.
- Longhorn requests remount shortly after.
- Kubernetes tries to delete `pod_replacement` in favor of `new_pod_replacement`.
- kubelet cannot clean up the broken mount for `pod_replacement`. It's a bit hazy why, but it seems that it realizes there is not a bind mount at `/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/<pv_name>/dev/<pod_uid>`, so instead of calling `NodeUnpublishVolume`, it tries to directly remove that file. It cannot, because it is bind mounted into the publish path (and is busy).

Long story short, if we create `replacement_pod` correctly (by ensuring it is staged before published), we can't hit this sequence of events.

File: csi/node_server.go [hunk 3]
```diff
@@ -852,3 +840,31 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
 
 	return nil, fmt.Errorf("failed to get mounter for volume %v unsupported volume capability %v", volume.Name, volumeCapability.GetAccessType())
 }
+
+// restageRequired determines whether it is necessary to manually call NodeUnstageVolume and NodeStageVolume again
+// before publishing. If it returns true, it may also return an error containing the underlying reason restaging is
+// required. restageRequired has side effects for v1 mount volumes due to its use of ensureMountPoint. These side
```
[the remark this answers is not in the document]
Comment (ejweber) [reply csi/node_server.go:846]: No. The side effects are specifically only for volumes with the mount access type (like https://github.com/container-storage-interface/spec/blob/886cd4877954568f9722cbe14d46b9fc16379d62/spec.md?plain=1#L977-L982). We do not call `ensureMountPoint` here for block volumes.

Please let me know if there's a different change I can make to the comment to clarify!

File: csi/node_server.go (hunk 2 above)
Comment (ejweber) [reply csi/node_server.go]: Yeah, the ` log.WithError(err).Warning(msg)` seemed reasonable. Adopted it in an amend.

**Review by PhanLe1010 — APPROVED** [review]
LGTM

File: csi/node_server.go [hunk 4]
```diff
@@ -852,3 +840,31 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
...
+// effects are neither harmful nor helpful, as ensureMountPoint will be called again in the restage flow.
+func restageRequired(volume *longhornclient.Volume,
+	volumeID, stagingTargetPath string,
+	mounter mount.Interface,
+	isBlock bool) (bool, error) {
+
+	if volume.DataEngine == string(longhorn.DataEngineTypeV2) {
+		return true, fmt.Errorf("always unstage v2 volume %v", volumeID)
+	}
```
Comment (derekbit) [inline csi/node_server.go:855]: Could you elaborate more on the v2 volume check? Thank you.

Comment (ejweber) [reply csi/node_server.go:855]: This logic is copied directly from lines 173-174 before the changes. I added the `true` return because of the way the function works now, but the previous logic made `unstageRequired` true for v2 volumes and logged this exact message. TBH, I don't know exactly why we need(ed) this behavior, but I don't think this PR should change it.

Comment (innobead) [reply csi/node_server.go:855]: If not required, we can remove it. Derek?

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
@mergify backport v1.6.x

**Comment by ejweber** [comment]
@mergify backport v1.5.x
